core/src/triggers.rs

```diff
@@ -8,82 +8,162 @@
 //! search trees (common lisp) or hash tables (racket) to quickly
 //! trigger hooks.
 
-use std::{cmp::min, ops::Deref};
+#![allow(clippy::expect_used, clippy::unwrap_in_result)]
 
-use dashmap::DashMap;
+use std::{cmp::min, sync::Arc};
+
+use dashmap::{DashMap, DashSet};
 use iroha_data_model::{
+    events::Filter as _,
     prelude::*,
     trigger::{self, Action, Repeats, Trigger},
 };
+use tokio::{sync::RwLock, task};
 
 use crate::smartcontracts::{self, FindError, InstructionType, MathError};
 
-/// Specialised structure that maps event filters to Triggers.
-#[derive(Debug, Default, Clone)]
-pub struct TriggerSet(
-    DashMap<trigger::Id, Action>, // TODO: Consider tree structures.
-);
+type Result<T> = std::result::Result<T, smartcontracts::Error>;
+
+/// Specialized structure that maps event filters to Triggers.
+/// TODO: trigger strong-typing
+#[derive(Debug, Default)]
+pub struct TriggerSet {
+    /// Triggers using [`DataEventFilter`]
+    data_triggers: DashMap<trigger::Id, Action<DataEventFilter>>,
+    /// Triggers using [`PipelineEventFilter`]
+    pipeline_triggers: DashMap<trigger::Id, Action<PipelineEventFilter>>,
+    /// Triggers using [`TimeEventFilter`]
+    time_triggers: DashMap<trigger::Id, Action<TimeEventFilter>>,
+    /// Triggers using [`ExecuteTriggerEventFilter`]
+    by_call_triggers: DashMap<trigger::Id, Action<ExecuteTriggerEventFilter>>,
+    /// Set of all ids. Used to check that every id is unique
+    ids: DashSet<trigger::Id>,
+    /// List of actions that should be triggered by events provided by `handle_*` methods.
+    /// Not being cloned
+    matched_ids: RwLock<Vec<(EventType, trigger::Id)>>,
+}
 
-impl From<TriggerSet> for Vec<trigger::Id> {
-    fn from(TriggerSet(map): TriggerSet) -> Self {
-        map.iter()
-            .map(|reference| reference.key().clone())
-            .collect()
+impl Clone for TriggerSet {
+    fn clone(&self) -> Self {
+        Self {
+            data_triggers: self.data_triggers.clone(),
+            pipeline_triggers: self.pipeline_triggers.clone(),
+            time_triggers: self.time_triggers.clone(),
+            by_call_triggers: self.by_call_triggers.clone(),
+            ids: self.ids.clone(),
+            matched_ids: RwLock::new(Vec::new()),
+        }
     }
 }
 
 impl TriggerSet {
-    /// Add another trigger to the [`TriggerSet`].
+    /// Add trigger with [`DataEventFilter`]
+    ///
+    /// # Errors
+    /// - If trigger with such id already exists
+    pub fn add_data_trigger(&self, trigger: Trigger<DataEventFilter>) -> Result<()> {
```

Comment:
Have you considered generics? I'm not 100% convinced they would work, but I think that a generic `add` method could help a lot with extra writing.